File: android/src/main/java/com/stallion/networkmanager/StallionSyncHandler.java
package com.stallion.networkmanager;

import android.util.Log;
import com.stallion.events.StallionEventManager;
import com.stallion.storage.StallionConfigConstants;
import com.stallion.storage.StallionMetaConstants;
import com.stallion.storage.StallionStateManager;
import com.stallion.storage.StallionConfig;
import com.stallion.utils.StallionSlotManager;
import com.stallion.events.StallionEventConstants.NativeProdEventTypes;

import org.json.JSONObject;
import java.util.concurrent.atomic.AtomicBoolean;

public class StallionSyncHandler {

  private static final String TAG = "StallionSyncHandler";
  private static final AtomicBoolean isSyncInProgress = new AtomicBoolean(false);
  private static final AtomicBoolean isDownloadInProgress = new AtomicBoolean(false);

  public static void sync() {
    // Ensure only one sync job runs at a time
    if (!isSyncInProgress.compareAndSet(false, true)) {
      return; // Exit if another job is already running
    }

    new Thread(() -> {
      try {
        // Fetch StallionStateManager and StallionConfig
        StallionStateManager stateManager = StallionStateManager.getInstance();
        StallionConfig config = stateManager.getStallionConfig();

        // Use appVersion directly from StallionConfig
        String appVersion = config.getAppVersion();
        String projectId = config.getProjectId();
        String appliedBundleHash = stateManager.stallionMeta.getActiveReleaseHash();

        // Prepare payload for API call
        JSONObject requestPayload = new JSONObject();
        requestPayload.put("appVersion", appVersion);
        requestPayload.put("platform", "android");
        requestPayload.put("projectId", projectId);
        requestPayload.put("appliedBundleHash", appliedBundleHash);

        // Make API call using StallionApiManager
        JSONObject releaseMeta = StallionApiManager.post(
          StallionApiConstants.STALLION_API_BASE + StallionApiConstants.STALLION_INFO_API_PATH,
          requestPayload.toString()
        );

        // Process API response
        processReleaseMeta(releaseMeta, appVersion);

      } catch (Exception e) {
        emitSyncError(e);
      } finally {
        // Reset the flag to allow new jobs
        isSyncInProgress.set(false);
      }
    }).start();
  }
  private static void processReleaseMeta(JSONObject releaseMeta, String appVersion) {
    Log.d(TAG, "🔄 Processing release meta for app version: " + appVersion);
    
    // Handle the current API response format from your server
    // The server returns: { updateAvailable: true, releaseHash: "...", downloadUrl: "...", ... }
    // instead of the expected: { success: true, data: { appliedBundleData: {...}, newBundleData: {...} } }
    
    boolean updateAvailable = releaseMeta.optBoolean("updateAvailable", false);
    Log.d(TAG, "📦 Update available: " + updateAvailable);
    
    if (updateAvailable) {
      // Convert current response format to expected format
      JSONObject newBundleData = new JSONObject();
      try {
        // Map your server's fields to what the existing code expects
        String downloadUrl = releaseMeta.optString("downloadUrl");
        String releaseHash = releaseMeta.optString("releaseHash");
        String targetAppVersion = releaseMeta.optString("targetAppVersion");
        long bundleSize = releaseMeta.optLong("bundleSize", 0);
        
        Log.d(TAG, "🔗 Download URL: " + downloadUrl);
        Log.d(TAG, "🔐 Release Hash: " + releaseHash);
        Log.d(TAG, "🎯 Target App Version: " + targetAppVersion);
        Log.d(TAG, "📏 Bundle Size: " + bundleSize + " bytes");
        
        newBundleData.put("downloadUrl", downloadUrl);
        newBundleData.put("checksum", releaseHash); // Your server uses "releaseHash", code expects "checksum"
        newBundleData.put("targetAppVersion", targetAppVersion);
        newBundleData.put("bundleSize", bundleSize); // Pass bundle size from API
        
        // Create appliedBundleData with default values (since your server doesn't provide this)
        JSONObject appliedBundleData = new JSONObject();
        appliedBundleData.put("isRolledBack", false); // Default to not rolled back
        appliedBundleData.put("targetAppVersion", appVersion);
        
        Log.d(TAG, "✅ Successfully adapted API response format");
        
        // Process using existing handlers
        handleAppliedReleaseData(appliedBundleData, appVersion);
        handleNewReleaseData(newBundleData);
        
      } catch (Exception e) {
        // If there's an error creating the adapted response, emit sync error
        Log.e(TAG, "❌ Failed to parse API response: " + e.getMessage());
        emitSyncError(new Exception("Failed to parse API response: " + e.getMessage()));
      }
    } else {
      Log.d(TAG, "📋 No update available, trying fallback format");
      // Fallback: try the original format for backwards compatibility
      if (releaseMeta.optBoolean("success")) {
        JSONObject data = releaseMeta.optJSONObject("data");
        if (data == null) {
          Log.d(TAG, "⚠️ No data object in response");
          return;
        }

        handleAppliedReleaseData(data.optJSONObject("appliedBundleData"), appVersion);
        handleNewReleaseData(data.optJSONObject("newBundleData"));
      } else {
        Log.d(TAG, "⚠️ No success field or updateAvailable field in response");
      }
    }
  }

  private static void handleAppliedReleaseData(JSONObject appliedData, String appVersion) {
    if (appliedData == null) return;

    boolean isRolledBack = appliedData.optBoolean("isRolledBack");
    String targetAppVersion = appliedData.optString("targetAppVersion");
    if (isRolledBack && appVersion.equals(targetAppVersion)) {
      StallionSlotManager.rollbackProd(false, "");
    }
  }

  private static void handleNewReleaseData(JSONObject newReleaseData) {
    if (newReleaseData == null) {
      Log.d(TAG, "⚠️ No new release data provided");
      return;
    }

    String newReleaseUrl = newReleaseData.optString("downloadUrl");
    String newReleaseHash = newReleaseData.optString("checksum");
    long bundleSize = newReleaseData.optLong("bundleSize", 0);

    Log.d(TAG, "📋 Handling new release data:");
    Log.d(TAG, "   URL: " + newReleaseUrl);
    Log.d(TAG, "   Hash: " + newReleaseHash);
    Log.d(TAG, "   Bundle Size: " + bundleSize + " bytes");

    StallionStateManager stateManager = StallionStateManager.getInstance();
    String lastRolledBackHash = stateManager.stallionMeta.getLastRolledBackHash();
    Log.d(TAG, "   Last rolled back hash: " + lastRolledBackHash);

    if (
        !newReleaseHash.isEmpty()
        && !newReleaseUrl.isEmpty()
        && !newReleaseHash.equals(lastRolledBackHash)
    ) {
      if(stateManager.getIsMounted()) {
        Log.d(TAG, "🚀 App is mounted, starting download immediately");
        downloadNewRelease(newReleaseHash, newReleaseUrl, bundleSize);
      } else {
        Log.d(TAG, "⏳ App not mounted, setting pending release");
        stateManager.setPendingRelease(newReleaseUrl, newReleaseHash);
      }
    } else {
      Log.d(TAG, "⏭️ Skipping download - hash empty, URL empty, or already rolled back");
    }
  }

  public static void downloadNewRelease(String newReleaseHash, String newReleaseUrl, long bundleSize) {
    // Ensure only one download job runs at a time
    if (!isDownloadInProgress.compareAndSet(false, true)) {
      return; // Exit if another job is already running
    }
    try {
      StallionStateManager stateManager = StallionStateManager.getInstance();
      StallionConfig config = stateManager.getStallionConfig();
      String downloadPath = config.getFilesDirectory()
        + StallionConfigConstants.PROD_DIRECTORY
        + StallionConfigConstants.TEMP_FOLDER_SLOT;
      String projectId = config.getProjectId();
      // Use the downloadUrl directly - it's already a complete presigned S3 URL
      String downloadUrl = newReleaseUrl;
      Log.d(TAG, "🔗 Using complete presigned URL directly: " + downloadUrl);
      Log.d(TAG, "📏 Bundle size from API: " + bundleSize + " bytes");

      long alreadyDownloaded = StallionDownloadCacheManager.getDownloadCache(config, downloadUrl, downloadPath);

      emitDownloadStarted(newReleaseHash, alreadyDownloaded > 0);

      // Use bundle size if available, otherwise fallback to original method
      if (bundleSize > 0) {
        Log.d(TAG, "📦 Using known bundle size for download");
        StallionFileDownloader.downloadBundleWithSize(
          downloadUrl,
          downloadPath,
          alreadyDownloaded,
          bundleSize,
        new StallionDownloadCallback() {
          @Override
          public void onReject(String prefix, String error) {
            isDownloadInProgress.set(false);
            emitDownloadError(newReleaseHash, prefix + error);
          }

          @Override
          public void onSuccess(String successPayload) {
            isDownloadInProgress.set(false);
            stateManager.stallionMeta.setCurrentProdSlot(StallionMetaConstants.SlotStates.NEW_SLOT);
            stateManager.stallionMeta.setProdTempHash(newReleaseHash);
            String currentProdNewHash = stateManager.stallionMeta.getProdNewHash();
            if(currentProdNewHash != null && !currentProdNewHash.isEmpty()) {
              StallionSlotManager.stabilizeProd();
            }
            stateManager.syncStallionMeta();
            StallionDownloadCacheManager.deleteDownloadCache(downloadPath);
            emitDownloadSuccess(newReleaseHash);
          }

          @Override
          public void onProgress(double downloadFraction) {
            // Optional: Handle progress updates
          }
        }
      );
      } else {
        Log.d(TAG, "📦 No bundle size available, using original download method with HEAD request");
        StallionFileDownloader.downloadBundle(
          downloadUrl,
          downloadPath,
          alreadyDownloaded,
          new StallionDownloadCallback() {
            @Override
            public void onReject(String prefix, String error) {
              isDownloadInProgress.set(false);
              emitDownloadError(newReleaseHash, prefix + error);
            }

            @Override
            public void onSuccess(String successPayload) {
              isDownloadInProgress.set(false);
              stateManager.stallionMeta.setCurrentProdSlot(StallionMetaConstants.SlotStates.NEW_SLOT);
              stateManager.stallionMeta.setProdTempHash(newReleaseHash);
              String currentProdNewHash = stateManager.stallionMeta.getProdNewHash();
              if(currentProdNewHash != null && !currentProdNewHash.isEmpty()) {
                StallionSlotManager.stabilizeProd();
              }
              stateManager.syncStallionMeta();
              StallionDownloadCacheManager.deleteDownloadCache(downloadPath);
              emitDownloadSuccess(newReleaseHash);
            }

            @Override
            public void onProgress(double downloadFraction) {
              // Optional: Handle progress updates
            }
          }
        );
      }
    } catch (Exception ignored) {
      isDownloadInProgress.set(false);
    }
  }

  private static void emitSyncError(Exception e) {
    JSONObject syncErrorPayload = new JSONObject();
    try {
      String syncErrorString = e.getMessage() != null ? e.getMessage() : "Unknown error";
      syncErrorPayload.put("meta", syncErrorString);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      NativeProdEventTypes.SYNC_ERROR_PROD.toString(),
      syncErrorPayload
    );
  }

  private static void emitDownloadError(String releaseHash, String error) {
    JSONObject errorPayload = new JSONObject();
    try {
      errorPayload.put("releaseHash", releaseHash);
      errorPayload.put("meta", error);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      NativeProdEventTypes.DOWNLOAD_ERROR_PROD.toString(),
      errorPayload
    );
  }

  private static void emitDownloadSuccess(String releaseHash) {
    JSONObject successPayload = new JSONObject();
    try {
      successPayload.put("releaseHash", releaseHash);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      NativeProdEventTypes.DOWNLOAD_COMPLETE_PROD.toString(),
      successPayload
    );
  }

  private static void emitDownloadStarted(String releaseHash, Boolean isResume) {
    JSONObject successPayload = new JSONObject();
    try {
      successPayload.put("releaseHash", releaseHash);
    } catch (Exception ignored) { }
    StallionEventManager.getInstance().sendEvent(
      isResume ? NativeProdEventTypes.DOWNLOAD_RESUME_PROD.toString(): NativeProdEventTypes.DOWNLOAD_STARTED_PROD.toString(),
      successPayload
    );
  }
}

File: ios/main/StallionSyncHandler.swift
//
//  StallionSyncHandler.swift
//  react-native-stallion
//
//  Created by Jasbir Singh Shergill on 29/01/25.
//

import Foundation

class StallionSyncHandler {

    private static var isSyncInProgress = false
    private static var isDownloadInProgress = false
    private static let syncQueue = DispatchQueue(label: "com.stallion.syncQueue")

    static func sync() {
        var shouldProceed = false

        syncQueue.sync {
            if !isSyncInProgress {
                isSyncInProgress = true
                shouldProceed = true
            }
        }

        guard shouldProceed else { 
            print("⚠️ Sync already in progress - skipping")
            return 
        }
        
        print("🚀 StallionSyncHandler.sync() started")
        
        // Send debug event to JavaScript
        let debugPayload: NSDictionary = ["message": "Native sync started"]
        Stallion.sendEventToRn(eventName: "SYNC_DEBUG", eventBody: debugPayload, shouldCache: false)

          DispatchQueue.global().async {
              do {
                  // Fetch StallionStateManager and StallionConfig
                  let stateManager = StallionStateManager.sharedInstance()
                  guard let config = stateManager?.stallionConfig else {
                      completeSync()
                      return
                  }

                  // Use appVersion directly from StallionConfig
                  let appVersion = config.appVersion ?? ""
                  let projectId = config.projectId ?? ""
                  let uid = config.uid ?? ""
                  let appliedBundleHash = stateManager?.stallionMeta?.getActiveReleaseHash() ?? ""

                  // Get environment from config via Info.plist
                  let environment = config.environment ?? "prod"
                  
                  // Prepare payload for API call
                  let requestPayload: [String: Any] = [
                      "appVersion": appVersion,
                      "platform": StallionConstants.PlatformValue,
                      "projectId": projectId,
                      "currentEnvironment": environment,
                      "appliedBundleHash": appliedBundleHash // Will be empty on first launch
                  ]

                  // Make API call using URLSession
                  makeApiCall(payload: requestPayload, appVersion: appVersion)

              } catch {
                  completeSync()
                  emitSyncError(error)
              }
          }
      }

      private static func makeApiCall(payload: [String: Any], appVersion: String) {
          guard let url = URL(string: StallionConstants.STALLION_API_BASE + StallionConstants.STALLION_INFO_API_PATH) else {
              emitSyncError(NSError(domain: "Invalid URL", code: -1))
              return
          }

          var request = URLRequest(url: url)
          request.httpMethod = "POST"
          request.setValue("application/json", forHTTPHeaderField: "Content-Type")

          // No authentication headers needed - API is public

          // Convert payload to JSON
          do {
              let jsonData = try JSONSerialization.data(withJSONObject: payload, options: [])
              request.httpBody = jsonData
          } catch {
              completeSync()
              emitSyncError(error)
              return
          }

          let task = URLSession.shared.dataTask(with: request) { data, response, error in
              if let error = error {
                  completeSync()
                  emitSyncError(error)
                  return
              }
            
              guard let data = data, let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                  completeSync()
                  let responseError = NSError(domain: "Invalid response from server", code: -2)
                  emitSyncError(responseError)
                  return
              }

              // Parse the JSON response
              print("📡 Raw API response data: \(String(data: data, encoding: .utf8) ?? "Unable to decode")")
              
              do {
                  if let releaseMeta = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                      print("✅ JSON parsed successfully: \(releaseMeta)")
                      
                      // Send debug event to JavaScript
                      let debugPayload: NSDictionary = ["message": "JSON parsed successfully", "updateAvailable": releaseMeta["updateAvailable"] ?? false]
                      Stallion.sendEventToRn(eventName: "SYNC_DEBUG", eventBody: debugPayload, shouldCache: false)
                      
                      completeSync()
                      processReleaseMeta(releaseMeta, appVersion: appVersion)
                  } else {
                      print("❌ Failed to parse JSON response")
                      completeSync()
                      let parsingError = NSError(domain: "Invalid JSON format", code: -3)
                      emitSyncError(parsingError)
                  }
              } catch {
                  print("❌ JSON parsing error: \(error)")
                  completeSync()
                  emitSyncError(error)
              }
          }

          task.resume()
      }
    
    private static func completeSync() {
        syncQueue.sync {
            isSyncInProgress = false
        }
    }
  
    private static func processReleaseMeta(_ releaseMeta: [String: Any], appVersion: String) {
        print("🔍 Processing release meta: \(releaseMeta)")
        
        // Check if update is available
        guard let updateAvailable = releaseMeta["updateAvailable"] as? Bool, updateAvailable else { 
            print("❌ No update available or updateAvailable key missing")
            return 
        }
        
        print("✅ Update available! Processing new release data...")
        // Process the release data directly from our API format
        handleNewReleaseData(releaseMeta)
    }

    private static func handleAppliedReleaseData(_ appliedData: [String: Any], appVersion: String) {
        guard let isRolledBack = appliedData["isRolledBack"] as? Bool,
              let targetAppVersion = appliedData["targetAppVersion"] as? String else { return }

        if isRolledBack && appVersion == targetAppVersion {
          StallionSlotManager.rollbackProd(withAutoRollback: false, errorString: "")
        }
    }

    private static func handleNewReleaseData(_ newReleaseData: [String: Any]) {
        print("🔍 Processing new release data: \(newReleaseData)")
        
        // Try our API format first
        var newReleaseUrl = newReleaseData["downloadUrl"] as? String
        var newReleaseHash = newReleaseData["releaseHash"] as? String
        
        // Fallback to legacy format
        if newReleaseHash == nil {
            newReleaseHash = newReleaseData["checksum"] as? String
        }
        
        guard let url = newReleaseUrl, let hash = newReleaseHash,
              !url.isEmpty, !hash.isEmpty else { 
            print("❌ Missing downloadUrl or release hash")
            print("downloadUrl: \(newReleaseData["downloadUrl"] ?? "nil")")
            print("releaseHash: \(newReleaseData["releaseHash"] ?? "nil")")
            print("checksum: \(newReleaseData["checksum"] ?? "nil")")
            return 
        }
        
        // Use the validated values
        let finalUrl = url
        let finalHash = hash
        
        print("✅ Found valid release data - URL: \(finalUrl)")
        print("✅ Release hash: \(finalHash)")

        let stateManager = StallionStateManager.sharedInstance()
        let lastRolledBackHash = stateManager?.stallionMeta?.lastRolledBackHash ?? ""

        if finalHash != lastRolledBackHash {
            if stateManager?.isMounted == true {
                print("✅ State manager is mounted - starting download...")
                downloadNewRelease(newReleaseHash: finalHash, newReleaseUrl: finalUrl)
            } else {
                print("⏳ State manager not mounted - storing pending release...")
                stateManager?.pendingReleaseUrl = finalUrl
                stateManager?.pendingReleaseHash = finalHash
            }
        } else {
            print("⚠️ Release hash matches last rolled back hash - skipping download")
        }
    }

    static func downloadNewRelease(newReleaseHash: String, newReleaseUrl: String) {
        print("🚀 Starting download for release hash: \(newReleaseHash)")
        print("🚀 Download URL: \(newReleaseUrl)")
        
        guard let stateManager = StallionStateManager.sharedInstance(),
              let config = stateManager.stallionConfig else { 
            print("❌ Failed to get state manager or config")
            return 
        }
      
        var shouldDownload = false
        syncQueue.sync {
            if !isDownloadInProgress {
                isDownloadInProgress = true
                shouldDownload = true
            }
        }
        guard shouldDownload else { return }

      let downloadPath = config.filesDirectory + "/" + StallionConstants.PROD_DIRECTORY + "/" + StallionConstants.TEMP_FOLDER_SLOT
      let projectId = config.projectId ?? ""
      
      // Use the download URL directly - it's already a complete presigned S3 URL
      guard let fromUrl = URL(string: newReleaseUrl) else { 
          print("❌ Invalid download URL: \(newReleaseUrl)")
          return 
      }
      
      print("✅ Using download URL: \(fromUrl)")

      emitDownloadStarted(releaseHash: newReleaseHash)

      StallionFileDownloader().downloadBundle(url: fromUrl, downloadDirectory: downloadPath, onProgress: { progress in
        // Handle progress updates if necessary
    }, resolve: { _ in
        completeDownload()
      stateManager.stallionMeta?.currentProdSlot =  SlotStates.newSlot
      stateManager.stallionMeta?.prodTempHash =  newReleaseHash
      if let currentProdNewHash = stateManager.stallionMeta?.prodNewHash,
         !currentProdNewHash.isEmpty {
          StallionSlotManager.stabilizeProd()
      }
      stateManager.syncStallionMeta()
      emitDownloadSuccess(releaseHash: newReleaseHash)
    }, reject: { code, prefix, error  in
        completeDownload()
      emitDownloadError(
        releaseHash: newReleaseHash,
        error: "\(String(describing: prefix))\(String(describing: error))"
      )
    })
  }
    
    private static func completeDownload() {
        syncQueue.sync {
            isDownloadInProgress = false
        }
    }

    // MARK: - Event Emission

    private static func emitSyncError(_ error: Error) {
      let syncErrorPayload: NSDictionary = ["meta": error.localizedDescription]
      Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesProd.SYNC_ERROR_PROD,
                             eventBody: syncErrorPayload,
                             shouldCache: true
      )
    }

    private static func emitDownloadError(releaseHash: String, error: String) {
        let errorPayload: NSDictionary = [
            "releaseHash": releaseHash,
            "meta": error
        ]
      Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesProd.DOWNLOAD_ERROR_PROD,
                             eventBody: errorPayload,
                             shouldCache: true
      )
    }

    private static func emitDownloadSuccess(releaseHash: String) {
      let successPayload: NSDictionary = ["releaseHash": releaseHash]
      Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesProd.DOWNLOAD_COMPLETE_PROD,
                             eventBody: successPayload,
                             shouldCache: true
      )
    }

    private static func emitDownloadStarted(releaseHash: String) {
        let startedPayload: NSDictionary = ["releaseHash": releaseHash]
        Stallion.sendEventToRn(eventName: StallionConstants.NativeEventTypesProd.DOWNLOAD_STARTED_PROD,
                             eventBody: startedPayload,
                             shouldCache: true
      )
    }
}

